### PR TITLE
`aspire update` command

### LINKED ISSUE #9135: `aspire update` command
We want to implement an `aspire update` command on the CLI. The goal of this command is to update the SDK and package references in an app host project as well as package references in referenced projects.

It should probably build upon any prior art that we have in the .NET SDK, but it will likely need to go further to support polyglot scenarios if we end up having packages for other languages/frameworks.

We probably want to have it be an interactive experience where the developer can run it locally and confirm each update as we suggest it (so they know what changed).

### PR DESCRIPTION
**Pull request by mitchdenny**
## Description

Fixes #9135

This PR introduces an `aspire update` command which can be used to update the SDK and package references for an Aspire app host and its referenced project.

~This PR is a DRAFT and is not ready to merge yet because it does not yet deal with the corner cases. But to test the functionality create an Aspire `9.3.1` project in a directory then install this build of the CLI and run `aspire update`. It should update everything Aspire related.~

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
- Did you add public API?
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [x] No
- Does the change require an update in our Aspire docs?
  - [x] Yes
    - Link to aspire-docs issue (consider using one of the following templates):
      - [New (or update) `doc-idea` template](https://github.com/dotnet/docs-aspire/issues/new?template=02-docs-request.yml)
      - [New `breaking-change` template](https://github.com/dotnet/docs-aspire/issues/new?template=04-breaking-change.yml)
      - [New `diagnostic` template](https://github.com/dotnet/docs-aspire/issues/new?template=06-diagnostic-addition.yml)
  - [ ] No

### CODE REVIEW

File: src/Aspire.Cli/Commands/UpdateCommand.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+            var projectFile = await _projectLocator.UseOrFindAppHostProjectFileAsync(null, cancellationToken);
+            var channels = await _packagingService.GetChannelsAsync(cancellationToken);
+
+            var channel = await _interactionService.PromptForSelectionAsync("Select a channel:", channels, (c) => c.Name, cancellationToken);
+
+            await _projectUpdater.UpdateProjectAsync(projectFile!, channel, cancellationToken);
+        }
+        catch (ProjectUpdaterException ex)
+        {
+            _interactionService.DisplayError(ex.Message);
```
Comment (davidfowl) [inline src/Aspire.Cli/Commands/UpdateCommand.cs]: Remind me, DisplayError escapes the input right?

**Comment by davidfowl** [comment]
Issues:

Move from daily to stable:

image

Packages that never had a stable release but are part of that channel don't work.

Updating from daily to default:

image

Package versions are the same?

**Comment by davidfowl** [comment]
Small nit: When you run update and there are multiple options, it asks which apphost you want to "run":

image

**Comment by davidfowl** [comment]
Ran this on eshop and got an exception:

image

```
🗄  Created settings file at '.aspire/settings.json'.
Unhandled exception: System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
   at Aspire.Cli.Projects.ProjectUpdater.<AnalyzeProjectAsync>d__15.MoveNext() + 0xb5c
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at Aspire.Cli.Projects.ProjectUpdater.<GetUpdateStepsAsync>d__8.MoveNext() + 0x208
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at Spectre.Console.Status.<>c__DisplayClass17_0`1.<<StartAsync>b__0>d.MoveNext() + 0x134
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at Spectre.Console.Progress.<>c__DisplayClass32_0`1.<<StartAsync>b__0>d.MoveNext() + 0x244
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at Spectre.Console.Internal.DefaultExclusivityMode.<RunAsync>d__3`1.MoveNext() + 0x1fc
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at Spectre.Console.Progress.<StartAsync>d__32`1.MoveNext() + 0x160
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at Spectre.Console.Status.<StartAsync>d__17`1.MoveNext() + 0x278
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at Aspire.Cli.Interaction.ConsoleInteractionService.<ShowStatusAsync>d__6`1.MoveNext() + 0x144
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at Aspire.Cli.Projects.ProjectUpdater.<UpdateProjectAsync>d__6.MoveNext() + 0x22c
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at Aspire.Cli.Commands.UpdateCommand.<ExecuteAsync>d__5.MoveNext() + 0x3c8
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at Aspire.Cli.Commands.BaseCommand.<>c__DisplayClass3_0.<<-ctor>b__0>d.MoveNext() + 0xd0
--- End of stack trace from previous location ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() + 0x24
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task) + 0x100
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task, ConfigureAwaitOptions) + 0x68
   at System.CommandLine.Invocation.InvocationPipeline.<InvokeAsync>d__0.MoveNext() + 0x4e8
```

**Comment by mitchdenny** [comment]
@copilot, on this branch I've added a bunch of new functionality. I want to start adding unit tests. The first thing that I want to add unit tests for is the `NuGetConfigMerger`. The purpose of the `NuGetConfigMerger` is to create a `NuGet.config` file in a specified directory based on the contents of the file already there (if it is there) and the mappings in a `PackageChannel` which are passed in.

Review the code and generate a unit test which tests this functionality. I will review the tests when they are done and provide further guidance. At this point just focus on writing the unit tests, don't change the code it is testing at all. If you encounter a scenario which is difficult to test, then stop and ask for guidance.

**Comment by mitchdenny** [comment]
@copilot, on this branch I've added a bunch of new functionality. I want to start adding unit tests. The first thing that I want to add unit tests for is the `NuGetConfigMerger`. The purpose of the `NuGetConfigMerger` is to create a `NuGet.config` file in a specified directory based on the contents of the file already there (if it is there) and the mappings in a `PackageChannel` which are passed in.
 
Review the code and generate a unit test which tests this functionality. I will review the tests when they are done and provide further guidance. At this point just focus on writing the unit tests, don't change the code it is testing at all. If you encounter a scenario which is difficult to test, then stop and ask for guidance.

**Comment by davidfowl** [comment]
More issues, if the SDK isn't resolvable the update command fails to work all up. Make the SDK `<Sdk Name="Aspire.AppHost.Sdk" Version="9.7.0" />` and try to run update, it'll fail.

**Comment by mitchdenny** [comment]
I think with eShop it's because of CPM. It's not working for CPM yet.

**Comment by davidfowl** [comment]
Can we improve the error handling there? Warn about CPM usage.

**Comment by mitchdenny** [comment]
> More issues, if the SDK isn't resolvable the update command fails to work all up. Make the SDK `<Sdk Name="Aspire.AppHost.Sdk" Version="9.7.0" />` and try to run update, it'll fail.

Yeah we rely on the project being usable by MSbuild to extract info.

**Comment by mitchdenny** [comment]
> More issues, if the SDK isn't resolvable the update command fails to work all up. Make the SDK `<Sdk Name="Aspire.AppHost.Sdk" Version="9.7.0" />` and try to run update, it'll fail.

Yeah we rely on the project being usable by MSbuild to extract info.

**Comment by davidfowl** [comment]
> Yeah we rely on the project being usable by MSbuild to extract info.

I think this is problematic and shouldn't break the command, it should fallback to another approach *OR* say we can't resolve the sdk version.

File: src/Aspire.Cli/Commands/UpdateCommand.cs (hunk 1 above)
Comment (mitchdenny) [reply src/Aspire.Cli/Commands/UpdateCommand.cs]: Explicitly added escapes for this just in case.

**Comment by mitchdenny** [comment]
> Small nit: When you run update and there are multiple options, it asks which apphost you want to "run":

image

**Comment by mitchdenny** [comment]
> > Yeah we rely on the project being usable by MSbuild to extract info.
> 
> I think this is problematic and shouldn't break the command, it should fallback to another approach _OR_ say we can't resolve the sdk version.

Done a pass over the locator logic to better identify projects that are unbuildable and did some centralization of the error logic across commands:

image

**Comment by mitchdenny** [comment]
> Can we improve the error handling there? Warn about CPM usage.

We now detect CPM and fail early:

image

**Comment by mitchdenny** [comment]
> Issues:
> 
> Move from daily to stable:
> 
> image
> Packages that never had a stable release but are part of that channel don't work.
> 
> Updating from daily to default:
> 
> image
> Package versions are the same?

Special cased scenario for stable channels where no stable package is present, in that case we allow for the resolution of prerelease packages:

image

**Comment by davidfowl** [comment]
> Done a pass over the locator logic to better identify projects that are unbuildable and did some centralization of the error logic across commands:

"builable" is not a great word. "Unable to resolve the SDK version from app host X".

I think this should still work and not break the upgrade flow (just skip the sdk). However, I'm ok with follow ups for this.

**Comment by davidfowl** [comment]
> Special cased scenario for stable channels where no stable package is present, in that case we allow for the resolution of prerelease packages:

I got this going from PR back to stable.

image

**Comment by mitchdenny** [comment]
> > Special cased scenario for stable channels where no stable package is present, in that case we allow for the resolution of prerelease packages:
> 
> I got this going from PR back to stable.
> 
> image

Could not repro on latest a build. I did fix a few bugs...

I've verified

default -> daily -> stable -> pr -> stable

**Comment by mitchdenny** [comment]
> > Done a pass over the locator logic to better identify projects that are unbuildable and did some centralization of the error logic across commands:
> 
> "builable" is not a great word. "Unable to resolve the SDK version from app host X".
> 
> I think this should still work and not break the upgrade flow (just skip the sdk). However, I'm ok with follow ups for this.

Agree that it's not a great term, I'll tweak that - but this error catches any scenario in which the MSBuild file might be unusable, including syntax errors in the XML (they all return the same exit code from MSBuild unfortunately). So, we'd have to step into doing error log sniffing to try and detect SDK errors specifically.

**Comment by mitchdenny** [comment]
This is what I have at the moment:

image

**Comment by mitchdenny** [comment]
OK, I think this is solid enough to merge in. Test coverage on ProjectUpdater is in the ~95% range is just some failure cases and some quirks of code coverage detection.

**Review by davidfowl — APPROVED** [review]
